### PR TITLE
Spinlockification

### PR DESCRIPTION
**Pull request by andyross**
Big refactoring to get the kernel internals using spinlocks pervasively.

The interesting patches are early in the series.  Later ones are mostly boilerplate changes.

A few spots have been identified as design points that might need new decisions in the future.  In almost all cases I picked "keep existing synchronization strategy in place" vs. "do the perfect thing".

### CODE REVIEW

File: kernel/sem.c [hunk 1]
```diff
@@ -31,6 +31,15 @@
 extern struct k_sem _k_sem_list_start[];
 extern struct k_sem _k_sem_list_end[];
 
+/* We use a system-wide lock to synchronize semaphores, which has
+ * unfortunate performance impact vs. using a per-object lock
+ * (semaphores are *very* widely used).  But per-object locks require
+ * significant extra RAM.  A properly spin-aware semaphore
+ * implementation would spin on atomic access to the count variable,
+ * and not a spinlock per se.  Useful optimization for the future...
```
Comment (pizi-nordic) [inline kernel/sem.c]: IMHO this calls for better atomic operations too...

File: kernel/include/ksched.h [hunk 2]
```diff
@@ -55,6 +55,22 @@ int _pend_curr(struct k_spinlock *lock, k_spinlock_key_t key,
 	       _wait_q_t *wait_q, s32_t timeout);
 int _reschedule(struct k_spinlock *lock, k_spinlock_key_t key);
 
+static inline int _pend_curr_unlocked(_wait_q_t *wait_q, s32_t timeout)
```
Comment (pizi-nordic) [inline kernel/include/ksched.h]: For me an ..._unlocked function which is taking the lock is a bit confusing.

File: kernel/include/kswap.h [hunk 3]
```diff
@@ -41,11 +42,15 @@ void _smp_release_global_lock(struct k_thread *thread);
  * Needed for SMP, where the scheduler requires spinlocking that we
  * don't want to have to do in per-architecture assembly.
  */
-static inline unsigned int _Swap(unsigned int key)
+static ALWAYS_INLINE unsigned int __Swap(unsigned int key,
+					 struct k_spinlock *lock,
```
[the remark this answers is not in the document]
Comment (andyross) [reply kernel/include/kswap.h]: The intent is that this code is going to be expanded macro-style into the (now) two different swap variants.  The is_spinlock is never going to exist at runtime.   Overloading the lock pointer (which does exist at runtime) with a compile-time constraint seems to me to be asking too much of the optimizer.

File: kernel/atomic_c.c [hunk 4]
```diff
@@ -254,15 +262,15 @@ atomic_val_t atomic_clear(atomic_t *target)
  */
 atomic_val_t atomic_or(atomic_t *target, atomic_val_t value)
```
[the remark this answers is not in the document]
Comment (andyross) [reply kernel/atomic_c.c]: We already do on architectures that support GCC atomic builtins, and a handful (xtensa, at least) that don't.  This is the fallback code for architectures without support.

**Comment by nashif** [comment]
@andyross what is the story with this one?

**Comment by andyross** [comment]
It's crazy stale, but probably updatable more easily than doing all the work from scratch.  Will resurrect.

**Comment by andyross** [comment]
Resurrected, rebased and in a spot or two rewritten.  There are a few known regressions I'm still chasing but this should be good enough for review if anyone has time.

Actually it's probably best it was done now after the spinlock validation feature last week.  There were a surprising number of recursive irq_lock calls lurking in hard-to-see places.

**Review by dbkinder — APPROVED** [review]
doc changes LGTM

**Comment by andyross** [comment]
Have it good to go on my rig.  Just the inevitable CI failure to squash.  Please review.

**Comment by andyross** [comment]
?!$!@#! I hate big refactoring series.  Bunch of surgery needed to tests/kernel/sched/scheduler_api to get it a minor stack size bump that doesn't blow past the budget on a dozen boards.  Now I think I have it all working...

**Review by nashif — APPROVED** [review]

**Comment by nashif** [comment]
@andyross please rebase

**Comment by andyross** [comment]
Rebased to cover the changes to userspace.c.  Note that I'm getting a sanitycheck yaml parse failure in samples/driver/counter on master at the moment.  Not sure if that's something weird with my tree or not, but doesn't seem related to this series.

**Comment by andyross** [comment]
A sudden stack overflow appeared (wasn't happening yesterday!).  Disabled that CMSIS sample on qemu_x86_64, as we can't change stack sizes there like we do with our own API.

**Pull request merged**